src/test/java/liqp/RenderSettingsTest.java

```diff
@@ -135,4 +146,96 @@ public void raiseExceptionsInStrictModeFalseTest2() {
         // Rendering should not terminate
         assertThat(rendered, is("FOO"));
     }
+
+    @Test
+    public void testEnvironmentMapConfigurator() throws Exception {
+        final String secretKey = getClass() + ".secretKey";
+
+        ParseSettings parseSettings = new ParseSettings.Builder().with(new Filter("secret") {
+            @Override
+            public Object apply(Object value, TemplateContext context, Object... params) {
+                ObjectAppender.Controller sb = context.newObjectAppender(3);
+                sb.append(value);
+                sb.append(" ");
+                sb.append(context.getEnvironmentMap().get(secretKey));
+                return sb.getResult();
+            }
+        }).build();
+
+        AtomicBoolean gotEnvironmentMap = new AtomicBoolean(false);
+        RenderSettings renderSettings = new RenderSettings.Builder().withEnvironmentMapConfigurator((
+            env) -> {
+            env.put(secretKey, "world");
+
+            gotEnvironmentMap.set(true);
+        }).build();
+
+        TemplateParser parser = new TemplateParser.Builder().withParseSettings(parseSettings)
+            .withRenderSettings(renderSettings).build();
+        Template template = parser.parse("{{ 'Hello' | secret }}");
+
+        assertFalse(gotEnvironmentMap.get());
+        assertEquals("Hello world", template.render());
+        assertTrue(gotEnvironmentMap.get());
+    }
+
+    @Test
+    public void testCustomRenderTransformer() throws Exception {
+        RenderSettings renderSettings = new RenderSettings.Builder().withRenderTransformer(
+            new CustomRenderTransformer()).build();
+
+        TemplateParser parser = new TemplateParser.Builder().withRenderSettings(renderSettings).build();
+        Template template = parser.parse("{{ 'Hello' }} {{ 'world' }}");
+
+        Object obj = template.prerender();
+        assertEquals(MyAppender.class, obj.getClass());
+
+        // our custom RenderTransformer allows access to appended fragments
+        List<Object> list = ((MyAppender) obj).getList();
+        assertEquals(Arrays.asList("Hello", " ", "world"), list);
+
+        // the final output for Template.render and Template.prerender.toString should be identical
+        assertEquals("Hello world", obj.toString());
+        assertEquals("Hello world", template.render());
+    }
+
+    private static final class MyAppender implements ObjectAppender.Controller {
+        final List<Object> list = new ArrayList<>();
+
+        @Override
+        public Object getResult() {
+            return this;
+        }
+
+        @Override
+        public void append(Object obj) {
+            list.add(obj);
+        }
+
+        public List<Object> getList() {
+            return list;
+        }
+
+        @Override
+        public String toString() {
```

Comment:
In this example, this `toString` method is called upon `render`, which means `prerender` really returns some intermediate representation. It's really different from `renderToObject` or similar.

You could also imagine that the step from `prerender` to `render` performs some additional sanity checks, like length checks, etc.